src/legacy/core_plugins/kibana/public/local_application_service/local_application_service.ts

```diff
@@ -0,0 +1,136 @@
+ */
+export class LocalApplicationService {
+  private apps: App[] = [];
+  private forwards: ForwardDefinition[] = [];
+  private idGenerator = htmlIdGenerator('kibanaAppLocalApp');
+
+  /**
+   * Register an app to be managed by the application service.
+   * This method works exactly as `core.application.register`.
+   *
+   * When an app is mounted, it is responsible for routing. The app
+   * won't be mounted again if the route changes within the prefix
+   * of the app (its id). It is fine to use whatever means for handling
+   * routing within the app.
+   *
+   * When switching to a URL outside of the current prefix, the app router
+   * shouldn't do anything because it doesn't own the routing anymore -
+   * the local application service takes over routing again,
+   * unmounts the current app and mounts the next app.
+   *
+   * @param app The app descriptor
+   */
+  register(app: App) {
+    this.apps.push(app);
+  }
+
+  /**
+   * Forwards every URL starting with `legacyAppId` to the same URL starting
+   * with `newAppId` - e.g. `/legacy/my/legacy/path?q=123` gets forwarded to
+   * `/newApp/my/legacy/path?q=123`.
+   *
+   * When setting the `keepPrefix` option, the new app id is simply prepended.
+   * The example above would become `/newApp/legacy/my/legacy/path?q=123`.
+   *
+   * This method can be used to provide backwards compatibility for URLs when
+   * renaming or nesting plugins. For route changes after the prefix, please
+   * use the routing mechanism of your app.
+   *
+   * @param legacyAppId The name of the old app to forward URLs from
+   * @param newAppId The name of the new app that handles the URLs now
+   * @param options Whether the prefix of the old app is kept to nest the legacy
+   * path into the new path
+   */
+  forwardApp(
+    legacyAppId: string,
+    newAppId: string,
+    options: { keepPrefix: boolean } = { keepPrefix: false }
+  ) {
+    this.forwards.push({ legacyAppId, newAppId, ...options });
+  }
+
+  /**
+   * Wires up listeners to handle mounting and unmounting of apps to
+   * the legacy angular route manager. Once all apps within the Kibana
+   * plugin are using the local route manager, this implementation can
+   * be switched to a more lightweight implementation.
+   *
+   * @param angularRouteManager The current `ui/routes` instance
+   */
+  apply(angularRouteManager: UIRoutes) {
+    this.apps.forEach(app => {
+      const wrapperElementId = this.idGenerator();
+      angularRouteManager.when(matchAllWithPrefix(app), {
+        outerAngularWrapperRoute: true,
+        reloadOnSearch: false,
+        reloadOnUrl: false,
+        template: `<div style="height:100%" id="${wrapperElementId}"></div>`,
+        controller($scope: IScope) {
+          const element = document.getElementById(wrapperElementId)!;
+          (async () => {
+            const onUnmount = await app.mount({ core: npStart.core }, { element, appBasePath: '' });
```

Comment:
I believe the `appBasePath` here shouldn't be empty. It probably should be the id of the current legacy app + the id of the app being mounted. For example, if I registered a app with id `dashboard`, I'd expect the appBasePath to be something like: `chrome.basePath.prepend('/kibana#dashboard')`

Reply:
Is this really helpful? AFAI understand the `appBasePath` is passed in so the application can make sure it's prefixing all links and routes correctly, right? So the contract is basically "if something changes after this prefix, it's up to you to handle it". If the base path switches technologies going from a path to a hash, this will require really weird code within the app to handle both cases (checking whether there is a hash in there and switching from history router to hash router or something like that).

I'm also not sure whether we can provide a meaningful "base path" here yet. Currently all consumers of the local application service have to use hash routing and make sure they handle the `app/kibana` part as well as the `#/dashboard` part of the url, so we have to do changes to the routing in this regard anyway when switching to the new platform and the real application router. The compat layer for legacy platform seems to follow the same principle: `src/legacy/ui/public/new_platform/new_platform.ts`

Reply:
@joshdover I'm going to merge this PR because it's blocking several other PRs to become reviewable and it's not actively used anywhere for now. If you still have concerns we can continue the discussion and adjust things in a separate PR.